### PR TITLE
assigning scalar to r-value tensor of arbitrary shape

### PR DESCRIPTION
**Pull request by ShubhamAgarwal12**
Added support for scalar assignment to tensor.

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3137)

### CODE REVIEW

File: cpp/open3d/core/Tensor.h [hunk 1]
```diff
@@ -147,21 +147,9 @@ class Tensor {
...
-            utility::LogError(
-                    "Assignment with scalar only works for scalar Tensor of "
-                    "shape ()");
-        }
-        DISPATCH_DTYPE_TO_TEMPLATE_WITH_BOOL(GetDtype(), [&]() {
-            scalar_t casted_v = static_cast<scalar_t>(v);
-            MemoryManager::MemcpyFromHost(GetDataPtr(), GetDevice(), &casted_v,
-                                          sizeof(scalar_t));
-        });
+        this->Fill(v);
```
Comment (yxlao) [inline cpp/open3d/core/Tensor.h:152]: Please add unit tests, including corner cases (type conversions, special shapes, etc.)

**Review by yxlao — CHANGES_REQUESTED** [review]

**Comment by yxlao** [comment]
@ShubhamAgarwal12 we'll need to merge master and resolve the conflicts

File: cpp/open3d/core/Tensor.h (hunk 1 above)
Comment (ShubhamAgarwal12) [reply cpp/open3d/core/Tensor.h:152]: Added

File: cpp/tests/core/Tensor.cpp [hunk 2]
```diff
@@ -2801,5 +2801,46 @@ TEST_P(TensorPermuteDevices, NumpyIO) {
...
+    t.Slice(0, 0, 1) = 255;
+    EXPECT_EQ(t.GetShape(), t1.GetShape());
+    EXPECT_EQ(t.ToFlatVector<uint8_t>(), t1.ToFlatVector<uint8_t>());
+
+    // Datatype implicit conversion with Slice.
+    t = core::Tensor::Init<bool>({false}, device);
+    t1 = core::Tensor::Init<bool>({true}, device);
+    t.Slice(0, 0, 1) = 1;
+    EXPECT_EQ(t.GetShape(), t1.GetShape());
+    EXPECT_EQ(t.ToFlatVector<bool>(), t1.ToFlatVector<bool>());
```
Comment (yxlao) [inline cpp/tests/core/Tensor.cpp:2842]: Please add test cases for tensor with one element (shape `{}`, `{1}`, `{1, 1}`) and without `Slice()`.

**Review by yxlao — CHANGES_REQUESTED** [review]

File: cpp/tests/core/Tensor.cpp (hunk 2 above)
Comment (ShubhamAgarwal12) [reply cpp/tests/core/Tensor.cpp:2863]: Done.

Comment (ShubhamAgarwal12) [reply cpp/tests/core/Tensor.cpp:2863]: Done

File: cpp/tests/core/Tensor.cpp [hunk 3]
```diff
@@ -2801,5 +2801,70 @@ TEST_P(TensorPermuteDevices, NumpyIO) {
...
 }
 
+TEST_P(TensorPermuteDevices, RValueScalar) {
+    const core::Device &device = GetParam();
+    core::Tensor t, t1;
+
+    // Assign scalars basic tests.
+    t = core::Tensor::Init<uint8_t>({{1, 2, 3, 4}, {5, 6, 7, 8}}, device);
+    t1 = core::Tensor::Init<uint8_t>({{1, 10, 3, 4}, {5, 6, 7, 8}}, device);
+    t[0][1] = 10;
```
Comment (yxlao) [inline cpp/tests/core/Tensor.cpp:2811]: If we look back to the `operator=` function that we modified, the purpose of this PR is to enable "assigning scalar value to a tensor of arbitrary shape", where the scalar value will broadcast to all elements in the tensor.  Previously it is only possible to assign scalar to a tensor with one element.

Therefore, we should test "assigning scalar value to a tensor of arbitrary shape", not assigning to a sliced tenor with only one element (e.g. `t[0][1]` has only one element, this is already supported before this PR).

I suggest re-writing all tests and do it more systematically
- First, we should test different shapes, e.g. create tensors with different shapes, e.g. `{}`, `{0}`, `{0, 1}`, `{1}`, `{1, 1}` `{1, 2}` and assign to the full tensor. 
- We shall also test dtype conversion, e.g. what if you assign a float value to an int tensor? Will it throw exception, or will it do the conversion automatically?

**Review by yxlao — CHANGES_REQUESTED** [review]

File: cpp/tests/core/Tensor.cpp (hunk 3 above)
Comment (ShubhamAgarwal12) [reply cpp/tests/core/Tensor.cpp]: Sure. Thanks for pointing this out. I've modified the tests with these changes and added few more.

**Review by yxlao — CHANGES_REQUESTED** [review]
The tests look much better now. Two additional minor changes and we're good to go.

> Reviewed 1 of 1 files at r1, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3137#-:-MW91OqO9WVIRDPorjIe:b40lrdj)* status: all files reviewed, 2 unresolved discussions (waiting on @ShubhamAgarwal12)

___
*[cpp/tests/core/Tensor.cpp, line 2806 at r6](https://reviewable.io/reviews/intel-isl/open3d/3137#-MW91-V-BDdPWxos7jAE:-MW91-V-BDdPWxos7jAF:bp6st1v) ([raw file](https://github.com/intel-isl/open3d/blob/4ed0120583ab09f631afd62ed19887ba8f5ff455/cpp/tests/core/Tensor.cpp#L2806)):*
> ```cpp
> TEST_P(TensorPermuteDevices, RValueScalar) {
>     const core::Device &device = GetParam();
>     core::Tensor t, t1;
> ```

rename `t1` to `t_ref`
___
*[cpp/tests/core/Tensor.cpp, line 2814 at r6](https://reviewable.io/reviews/intel-isl/open3d/3137#-MW912pK9RvxoD_l06Y2:-MW912pK9RvxoD_l06Y3:b-fpo2ea) ([raw file](https://github.com/intel-isl/open3d/blob/4ed0120583ab09f631afd62ed19887ba8f5ff455/cpp/tests/core/Tensor.cpp#L2814)):*
> ```cpp
>     EXPECT_EQ(t.GetShape(), t1.GetShape());
>     EXPECT_EQ(t.GetDtype(), t1.GetDtype());
>     EXPECT_EQ(t.ToFlatVector<int32_t>(), t1.ToFlatVector<int32_t>());
> ```

use `Tensor::AllClose()` to compare tensor with another tensor, e.g. 
```
EXPECT_TRUE(t.AllClose(t_ref));
```

**Comment by yxlao** [comment]
Also, the title of this PR should be updated, we are not assigning to an `lvalue`. We have to convert a tensor to `rvalue` and then assign a scalar to it. 

This is actually `rvalue_tensor = rvalue_scalar` assignment, where `rvalue_tensor` can be of arbitrary shape.

**Review by ShubhamAgarwal12 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3137)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @yxlao)

___
*[cpp/tests/core/Tensor.cpp, line 2806 at r6](https://reviewable.io/reviews/intel-isl/open3d/3137#-MW91-V-BDdPWxos7jAE:-MW9I09T6EzJ2xutZeoN:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/fef7c38bdfdddbb0a10138867e47c121a7fe7d4f/cpp/tests/core/Tensor.cpp#L2806)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

rename `t1` to `t_ref`
</blockquote></details>

Done.
___
*[cpp/tests/core/Tensor.cpp, line 2814 at r6](https://reviewable.io/reviews/intel-isl/open3d/3137#-MW912pK9RvxoD_l06Y2:-MW9I0Wj7d2dQk2p9P-a:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/fef7c38bdfdddbb0a10138867e47c121a7fe7d4f/cpp/tests/core/Tensor.cpp#L2814)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

use `Tensor::AllClose()` to compare tensor with another tensor, e.g. 
```
EXPECT_TRUE(t.AllClose(t_ref));
```
</blockquote></details>

Done.

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3137)* status: 1 of 2 files reviewed, 5 unresolved discussions (waiting on @ShubhamAgarwal12 and @yxlao)

___
*[cpp/open3d/core/Tensor.h, line 148 at r7](https://reviewable.io/reviews/intel-isl/open3d/3137#-MW9IQxBChD4xWayCUYn:-MW9HoDmCncS2O-ahhJP:b-iinnxa) ([raw file](https://github.com/intel-isl/open3d/blob/0fdf84fbb8ef08bfe0135c29e27fc647dc1bcee0/cpp/open3d/core/Tensor.h#L148)):*
> ```objc
>     Tensor& operator=(Tensor&& other) &&;
> 
>     /// Tensor assignment rvalue = rvalue_scalar, e.g. `tensor_a[0] = 100`
> ```

change `rvalue_scalar` to `scalar`
___
*[cpp/open3d/core/Tensor.h, line 151 at r7](https://reviewable.io/reviews/intel-isl/open3d/3137#-MW9IT30-vk8RgfH2M-T:-MW9IT30-vk8RgfH2M-U:b-jy4itw) ([raw file](https://github.com/intel-isl/open3d/blob/0fdf84fbb8ef08bfe0135c29e27fc647dc1bcee0/cpp/open3d/core/Tensor.h#L151)):*
> ```objc
>     /// Implicit casting is performed to the underlying dtype.
>     template <typename T>
>     Tensor& operator=(const T& v) && {
> ```

change signature to `Tensor& operator=(const T v) &&`
___
*[cpp/open3d/core/Tensor.h, line 151 at r7](https://reviewable.io/reviews/intel-isl/open3d/3137#-MW9Io2o3AmPF9-VKHd0:-MW9Io2o3AmPF9-VKHd1:b-1vczj5) ([raw file](https://github.com/intel-isl/open3d/blob/f07150d73b3161df893e6a81b88207e7c1148387/cpp/open3d/core/Tensor.h#L151)):*
> ```objc
>     /// Implicit casting is performed to the underlying dtype.
>     ///
>     /// Note that we don't have lvalue = rvalue_scalar, e.g. we don't support
> ```

don't remove this comment, this is still true (could you verify it?)

change `rvalue_scalar` to `scalar`

**Review by ShubhamAgarwal12 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3137)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @yxlao)

___
*[cpp/open3d/core/Tensor.h, line 148 at r7](https://reviewable.io/reviews/intel-isl/open3d/3137#-MW9IQxBChD4xWayCUYn:-MW9RLo65T3MYSaqqk_8:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/0fdf84fbb8ef08bfe0135c29e27fc647dc1bcee0/cpp/open3d/core/Tensor.h#L148)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

change `rvalue_scalar` to `scalar`
</blockquote></details>

Done.
___
*[cpp/open3d/core/Tensor.h, line 151 at r7](https://reviewable.io/reviews/intel-isl/open3d/3137#-MW9Io2o3AmPF9-VKHd0:-MW9RBOQCbdFdKBCrzXM:b-n78lwm) ([raw file](https://github.com/intel-isl/open3d/blob/f07150d73b3161df893e6a81b88207e7c1148387/cpp/open3d/core/Tensor.h#L151)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

don't remove this comment, this is still true (could you verify it?)

change `rvalue_scalar` to `scalar`


</blockquote></details>

Yes. This is true. We need to use t.AsRvalue().
 
___
*[cpp/open3d/core/Tensor.h, line 151 at r7](https://reviewable.io/reviews/intel-isl/open3d/3137#-MW9IT30-vk8RgfH2M-T:-MW9R_hy5gbxYUbuTMGY:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/0fdf84fbb8ef08bfe0135c29e27fc647dc1bcee0/cpp/open3d/core/Tensor.h#L151)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

change signature to `Tensor& operator=(const T v) &&`
</blockquote></details>

Done.

**Review by yxlao — APPROVED** [review]
> Reviewed 1 of 1 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3137)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ShubhamAgarwal12)
